unified_executor/include/execution/shape.hpp

```diff
@@ -5,14 +5,27 @@
 #include <utility>
 #include <vector>
 
+
+template <typename Shape, unsigned _size = Shape::dim()>
+struct shape_wrapper_t {
+  shape_wrapper_t() = default;
+
+  unsigned get(unsigned pos) const { return Shape::sizes[pos]; }
+
+  unsigned operator[](unsigned dim) { return get(dim); }
+};
+
 template <unsigned _s0> constexpr unsigned prod() { return _s0; }
 
 template <unsigned _s0, unsigned _s1, unsigned... _sizes>
 constexpr unsigned prod() {
   return prod<_s1, _sizes...>() * _s0;
 };
 
-template <unsigned _s0 = 0, unsigned... _sizes> struct shape_t {
+template <unsigned _s0 = 0, unsigned... _sizes> struct shape_t: shape_wrapper_t<shape_t<_s0, _sizes...>, 1 + sizeof...(_sizes)> {
```

Comment:
This is weird. `shape_t` depends on `shape_wrapper_t` which depends on `shape_t` with the same signature. This seems problematic.

Reply:
```suggestion
template <typename <template>Shape, unsigned _s0, unsigned ..._sizes, unsigned _size = Shape::dim()>
struct shape_wrapper_t 
```

```suggestion
template <unsigned _s0 = 0, unsigned... _sizes> struct shape_t: shape_wrapper_t<shape_t, _s0, _sizes..., 1 + sizeof...(_sizes)> 
```

This would resolve this issue I believe.

Reply:
1. `shape_t` doesn't always fit `template <typename>Shape`
2. `Shape` is a template and thus `::dim()` isn't available without the parameters

I might be wrong, not a C++ compiler :D